### DISCUSSION TITLE
Saving timestamps if a segment is lost

### DISCUSSION THREAD

**Discussion by HellbringerOnline** [streamlink/streamlink#4617] [Q&A] [answered]
Hello. I'm trying to save twitch streams using a streamlink because some streamers use two audio tracks and the music is not recorded in the video.
Then I download the video and add the streamlink audio track to the video.
If some segment of the stream is lost, then it is discarded and the next live segment is inserted in its place. But this affects the overall duration of the track and I end up getting desynchronized from where the segment got lost.
How can I save timestamps and replace lost segments with silence and blank video with segment duration?
![image](https://user-images.githubusercontent.com/14059683/174822639-ba95e6d9-06a5-446f-b6e9-7f69b782edaf.png)

**Comment by bastimeyer** [comment, marked as answer]
> If some segment of the stream is lost, then it is discarded and the next live segment is inserted in its place.

What do you mean with "is lost"?

HLS segment download failures can be adjusted with these options:

- https://streamlink.github.io/cli.html#cmdoption-stream-segment-attempts
- https://streamlink.github.io/cli.html#cmdoption-stream-segment-timeout

If you're however talking about segment filtering of ads via the Twitch plugin's `--twitch-disable-ads`, then you're out of luck, because Twitch simply doesn't make the missing segments available to you, and what Streamlink does in this case is avoiding stream discontinuities from the replacement segments, aka the ads.

> How can I save timestamps

You can't. Stream discontinuities are unsupported by Streamlink (#4535), and this also applies to stream data that gets written to the file system. Streamlink's main intention is piping live streams to a video player, so it always has to output a single progressive stream. If there are discontinuities in the output, either due to segment download failures, ad insertions or the filtering of those, then the output will have gaps. Writing stream data to a new file after a discontinuity is not supported.

**Reply by HellbringerOnline** [reply]
> What do you mean with "is lost"?

I mean the complete loss of the segment due to problems with the connection.

> You can't. Stream discontinuities are unsupported by Streamlink (#4535), and this also applies to stream data that gets written to the file system. Streamlink's main intention is piping live streams to a video player, so it always has to output a single progressive stream. If there are discontinuities in the output, either due to segment download failures, ad insertions or the filtering of those, then the output will have gaps. Writing stream data to a new file after a discontinuity is not supported.

I'm not talking about replacing the lost segment with a generated segment with silence and blank video, but no data on the timeline, which the player will perceive as pausing the video with no sound but continuing to increase the played time counter. I hope I explained clearly.

**Reply by bastimeyer** [reply]
> I mean the complete loss of the segment due to problems with the connection.

> https://streamlink.github.io/cli.html#cmdoption-stream-segment-attempts
> https://streamlink.github.io/cli.html#cmdoption-stream-segment-timeout

----

> I'm not talking about replacing the lost segment with a generated segment with silence and blank video, but no data on the timeline, which the player will perceive as pausing the video with no sound but continuing to increase the played time counter. I hope I explained clearly.

This is not how it works.

You can try re-encoding the audio track with FFmpeg after you've finished downloading the stream, because the presentation timestamps should remain intact after a discontinuity, and FFmpeg might be able to fix the audio track, so that you'll have the right length. That isn't something that Streamlink can do.

**Reply by HellbringerOnline** [reply]
> You can try re-encoding the audio track with FFmpeg after you've finished downloading the stream, because the presentation timestamps should remain intact after a discontinuity, and FFmpeg might be able to fix the audio track, so that you'll have the right length. That isn't something that Streamlink can do.

The problem is that i was trying to extract the audio track from the corrupted file without loss and ffmpeg is dropping the corrupted packets, now i can see it. Thanks i will try to fill broken packets with silence via ffmpeg.
